src/bin/pg_autoctl/supervisor.c

```diff
@@ -612,9 +969,139 @@ supervisor_shutdown_sequence(Supervisor *supervisor)
 }
 
 
+/*
+ * supervisor_dynamic_handle calls the dynamic handler if exits. It is the
+ * handler's responsibility to add or remove dynamic services using the provided
+ * api. The last argument, passed from our caller down to the dynamicHandler
+ * function, will contain the net difference of succefully added and removed
+ * services.
+ *
+ * Returns void.
+ */
+static void
+supervisor_dynamic_handle(Supervisor *supervisor, int *diffCount)
+{
+	/*
+	 * Call the dynamic function if defined. The last argument, diffCount
+	 * contains the of count of services affected, e.g. 1 started => 1,
+	 * 1 stopped => -1, 2 started and 3 stopped => -1 etc
+	 */
+	if (supervisor->dynamicHandler)
+	{
+		supervisor->dynamicHandler(supervisor, supervisor->dynamicHandlerArg,
+								   diffCount);
+	}
+
+	asked_to_handle_dynamic = 0;
+}
+
+
+/*
+ * supervisor_handle_failed_restart decides what action to take when a service
+ * has failed to restart. Currently one of the three can happen:
+ *	* continue running the rest of the services
+ *	* exit the whole program with a happy code
+ *	* exit the whole program with failed code
+ *
+ * The action is depended on whether the service is held in the dynamic array.
```

Comment:
The restart policy should be the same regardless. For some of the use cases I know off, the service will be required to be PERSISTANT, otherwise the feature is not useful.

The disconnect is not about the restart policy, it is about the everything failing if one service fails.

Users not being able to connect to one pgbouncer does not mean that there are not others to connect to. Rights?

Reply:
Our inspiration for the supervisor implementation is the [Erlang OTP supervisor module](https://erlang.org/doc/man/supervisor.html) and I think what you're trying to implement is the supervisor strategy (one_for_one, one_for_all, rest_for_one, etc). Though I am not sure this is exactly what you have in mind.

I would like us to come up with a strategy that we can easily change later, either because the next dynamic service after pgbouncer is going to require another way to do it, or because in the discussion we realise we need to change our mind on how pgbouncer failures to run are to be handled.